pkg/provision/gce.go

```diff
@@ -0,0 +1,210 @@
+		Name:         host.Name,
+		MachineType:  fmt.Sprintf("zones/%s/machineTypes/%s", host.Additional["zone"], host.Plan),
+		CanIpForward: true,
+		Zone:         fmt.Sprintf("projects/%s/zones/us-central1-a", host.Additional["projectid"]),
+		Disks: []*compute.AttachedDisk{
+			{
+				AutoDelete: true,
+				Boot:       true,
+				DeviceName: host.Name,
+				Mode:       "READ_WRITE",
+				Type:       "PERSISTENT",
+				InitializeParams: &compute.AttachedDiskInitializeParams{
+					Description: "Inlets-Operator exit node",
+					DiskName:    host.Name,
+					DiskSizeGb:  10,
+					SourceImage: host.OS,
+				},
+			},
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "startup-script",
+					Value: &host.UserData,
+				},
+			},
+		},
+		Tags: &compute.Tags{
+			Items: []string{"http-server", "https-server", "inlets"},
+		},
+		Scheduling: &compute.Scheduling{
+			AutomaticRestart:  &autoRestart,
+			OnHostMaintenance: "MIGRATE",
+			Preemptible:       false,
+		},
+		NetworkInterfaces: []*compute.NetworkInterface{
+			{
+				AccessConfigs: []*compute.AccessConfig{
+					{
+						Type: "ONE_TO_ONE_NAT",
+						Name: "External NAT",
+					},
+				},
+				Network: "global/networks/default",
+			},
+		},
+		ServiceAccounts: []*compute.ServiceAccount{
+			{
+				Email: "default",
+				Scopes: []string{
+					compute.ComputeScope,
+				},
+			},
+		},
+	}
+
+	op, err := gce.gceProvisioner.Instances.Insert(host.Additional["projectid"], host.Additional["zone"], instance).Do()
+	if err != nil {
+		return nil, fmt.Errorf("could not provision GCE instance: %v", err)
+	}
+
+	status := ""
+
+	if op.Status == "RUNNING" {
+		status = ActiveStatus
+	}
+	return &ProvisionedHost{
+		ID:     constructCustomGCEID(host.Name, host.Additional["zone"], host.Additional["projectid"]),
+		Status: status,
+	}, nil
+
+}
+
+// CheckInletsFirewallRuleExists checks if the inlets firewall rule exists or not
+func (gce *GCEProvisioner) CheckInletsFirewallRuleExists(id string, inletsPort string) (bool, error) {
+	_, _, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return false, fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+	firewallRuleName := "inlets"
+	op, err := gce.gceProvisioner.Firewalls.Get(projectID, firewallRuleName).Do()
+	if err != nil {
+		return false, fmt.Errorf("Could not get inlets firewall rule: %v", err)
+	}
+	if op.Name == firewallRuleName {
+		for _, firewallRule := range op.Allowed {
+			for _, port := range firewallRule.Ports {
+				if port == inletsPort {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// CreateInletsFirewallRule creates a firewall rule opening up the control port for inlets
+func (gce *GCEProvisioner) CreateInletsFirewallRule(id string, inletsPort string) error {
+	_, _, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+	firewallRuleName := "inlets"
```

Comment:
What happens if this firewall rule is created more than once?

Reply:
It would give an error. Should I delete it with the deletion of the instance or ignore the error?

Reply:
Actually we are already checking for if the firewall Rule already exists in controller.go. Is it still possible that the firewall rule can be created more than once. I am moving the code to the gce.go file anyways :)

Reply:
I tried deleting the firewall rule when the exit node is deleted. But I guess this won't be a good idea considering it is possible for the user to expose multiple services, in-turn creating multiple exit-nodes, even if a single svc is deleted, it will cause the client to lose access to all of the exit-nodes, as they are targeted with a single firewall rule with a network-tag on the instances - "inlets"